### PR TITLE
Basic profiling tools for distributions

### PR DESCRIPTION
**Pull request by neerajprad**
Adds some basic profiling tools for distributions. See #460.

 - We can add an `infer.py` script which can be invoked in a similar manner. 
 - The common profiling utilities will be in `profiling_utils.py`. For instance the `@Profile` annotation can be applied to any function, which when run returns the function's value along with the side effect from the profiler (mostly terminal output, but can also be the profiler dump in `/data`).

```
@Profile(
    tool='timeit',
    tool_cfg={'repeat': 5},
    fn_id=lambda dist, batch_size, *args, **kwargs: 'sample_' + dist.dist_class.__name__)
def sample(dist, batch_size, *args, **kwargs):
    return dist.sample(batch_size=batch_size, *args, **kwargs)
```
 - There is a context manager which serves as an output buffer for running profiling tasks after which they can all be printed together using a tabular format. See examples below.

*Some example invocations:*

`python -m profiler.distributions --dist bernoulli normal dirichlet --tool timeit` 
<details>
<summary> output </summary>

```
+--------------+----------+---------------+
| DISTRIBUTION | SAMPLE   | BATCH_LOG_PDF |
+--------------+----------+---------------+
| bernoulli    | 0.004779 | 0.004176      |
+--------------+----------+---------------+
| normal       | 0.014364 | 0.012648      |
+--------------+----------+---------------+
| dirichlet    | 0.993340 | 0.062922      |
+--------------+----------+---------------+
```
</details>

`python -m profiler.distributions --dist --tool cprofile` 

<details>
<summary> output </summary>

```
+---------------+----------------------------------------------------------------------------------+
| DISTRIBUTION  | bernoulli                                                                        |
+---------------+----------------------------------------------------------------------------------+
| SAMPLE        | Thu Nov  9 19:10:20 2017                                                         |
|               | /Users/npradhan/workspace/pyro_dev/pyro/profiler/data/sample_Bernoulli           |
|               |                                                                                  |
|               |          73 function calls in 0.006 seconds                                      |
|               |                                                                                  |
|               |    Ordered by: cumulative time                                                   |
|               |    List reduced from 44 to 22 due to restriction <0.5>                           |
|               |                                                                                  |
|               |    ncalls  tottime  percall  cumtime  percall filename:lineno(function)          |
|               |         1    0.000    0.000    0.006    0.006 distributions.py:74(sample)        |
|               |         1    0.000    0.000    0.006    0.006 random_primitive.py:36(sample)     |
|               |         1    0.000    0.000    0.006    0.006 bernoulli.py:72(sample)            |
|               |         1    0.006    0.006    0.006    0.006 {torch._C.bernoulli}               |
|               |         1    0.000    0.000    0.000    0.000 bernoulli.py:33(__init__)          |
|               |         1    0.000    0.000    0.000    0.000 util.py:150(get_probs_and_logits)  |
|               |         7    0.000    0.000    0.000    0.000 {built-in method apply}            |
|               |         1    0.000    0.000    0.000    0.000 variable.py:428(clamp)             |
|               |         1    0.000    0.000    0.000    0.000 pointwise.py:144(forward)          |
|               |         1    0.000    0.000    0.000    0.000 variable.py:819(__sub__)           |
|               |         1    0.000    0.000    0.000    0.000 variable.py:331(sub)               |
|               |         1    0.000    0.000    0.000    0.000 variable.py:324(_sub)              |
|               |         2    0.000    0.000    0.000    0.000 variable.py:721(expand)            |
|               |         1    0.000    0.000    0.000    0.000 basic_ops.py:26(forward)           |
|               |         2    0.000    0.000    0.000    0.000 tensor.py:107(forward)             |
|               |         1    0.000    0.000    0.000    0.000 {torch._C.log}                     |
|               |         1    0.000    0.000    0.000    0.000 tensor.py:308(__mul__)             |
|               |         1    0.000    0.000    0.000    0.000 {method 'mul' of                   |
|               | 'torch._C.ByteTensorBase' objects}                                               |
|               |         1    0.000    0.000    0.000    0.000 {method 'sub' of                   |
|               | 'torch._C.FloatTensorBase' objects}                                              |
|               |         1    0.000    0.000    0.000    0.000 variable.py:374(log)               |
|               |         1    0.000    0.000    0.000    0.000 {method 'ge' of                    |
|               | 'torch._C.FloatTensorBase' objects}                                              |
|               |         1    0.000    0.000    0.000    0.000 {torch._C.log1p}                   |
|               |                                                                                  |
|               |                                                                                  |
|               |                                                                                  |
+---------------+----------------------------------------------------------------------------------+
| BATCH_LOG_PDF | Thu Nov  9 19:10:20 2017                                                         |
|               | /Users/npradhan/workspace/pyro_dev/pyro/profiler/data/batch_log_pdf_Bernoulli    |
|               |                                                                                  |
|               |          173 function calls in 0.006 seconds                                     |
|               |                                                                                  |
|               |    Ordered by: cumulative time                                                   |
|               |    List reduced from 67 to 34 due to restriction <0.5>                           |
|               |                                                                                  |
|               |    ncalls  tottime  percall  cumtime  percall filename:lineno(function)          |
|               |         1    0.000    0.000    0.006    0.006 distributions.py:82(batch_log_pdf) |
|               |         1    0.000    0.000    0.006    0.006                                    |
|               | random_primitive.py:44(batch_log_pdf)                                            |
|               |         1    0.000    0.000    0.006    0.006 bernoulli.py:78(batch_log_pdf)     |
|               |        22    0.000    0.000    0.006    0.000 {built-in method apply}            |
|               |         3    0.000    0.000    0.002    0.001 variable.py:812(__add__)           |
|               |         3    0.000    0.000    0.002    0.001 variable.py:318(add)               |
|               |         3    0.000    0.000    0.002    0.001 variable.py:311(_add)              |
|               |         3    0.000    0.000    0.002    0.001 basic_ops.py:9(forward)            |
|               |         3    0.002    0.001    0.002    0.001 {method 'add' of                   |
|               | 'torch._C.FloatTensorBase' objects}                                              |
|               |         1    0.000    0.000    0.001    0.001 variable.py:828(__mul__)           |
|               |         1    0.000    0.000    0.001    0.001 variable.py:337(mul)               |
|               |         1    0.000    0.000    0.001    0.001 basic_ops.py:43(forward)           |
|               |         1    0.001    0.001    0.001    0.001 {method 'mul' of                   |
|               | 'torch._C.FloatTensorBase' objects}                                              |
|               |         3    0.000    0.000    0.001    0.000 variable.py:819(__sub__)           |
|               |         3    0.000    0.000    0.001    0.000 variable.py:331(sub)               |
|               |         3    0.000    0.000    0.001    0.000 variable.py:324(_sub)              |
|               |         3    0.000    0.000    0.001    0.000 basic_ops.py:26(forward)           |
|               |         3    0.001    0.000    0.001    0.000 {method 'sub' of                   |
|               | 'torch._C.FloatTensorBase' objects}                                              |
|               |         1    0.000    0.000    0.001    0.001 {torch._C.sum}                     |
|               |         1    0.000    0.000    0.001    0.001 variable.py:475(sum)               |
|               |         1    0.000    0.000    0.001    0.001 reduce.py:10(forward)              |
|               |         1    0.001    0.001    0.001    0.001 {method 'sum' of                   |
|               | 'torch._C.FloatTensorBase' objects}                                              |
|               |         5    0.000    0.000    0.000    0.000 variable.py:860(__neg__)           |
|               |         5    0.000    0.000    0.000    0.000 basic_ops.py:218(forward)          |
|               |         5    0.000    0.000    0.000    0.000 {method 'neg' of                   |
|               | 'torch._C.FloatTensorBase' objects}                                              |
|               |         1    0.000    0.000    0.000    0.000 bernoulli.py:33(__init__)          |
|               |         1    0.000    0.000    0.000    0.000 util.py:150(get_probs_and_logits)  |
|               |         2    0.000    0.000    0.000    0.000 variable.py:428(clamp)             |
|               |         1    0.000    0.000    0.000    0.000 bernoulli.py:46(batch_shape)       |
|               |         1    0.000    0.000    0.000    0.000 pointwise.py:144(forward)          |
|               |         1    0.000    0.000    0.000    0.000 variable.py:724(expand_as)         |
|               |         2    0.000    0.000    0.000    0.000 variable.py:368(exp)               |
|               |         1    0.000    0.000    0.000    0.000 pointwise.py:295(forward)          |
|               |         2    0.000    0.000    0.000    0.000 variable.py:374(log)               |
|               |                                                                                  |
|               |                                                                                  |
|               |                                                                                  |
+---------------+----------------------------------------------------------------------------------+
```
</details>

This also dumps the output from the profiler in the `data` directory which can then be visualized using `snakeviz`.

*TODO:*
 - Add support for more tools - particularly for memory profiling.
 - Finetune output of cProfile for better readability.
 - Add some model/guide pairs in `profiler.infer` on which we can run SVI for profiling. Maybe reuse @martinjankowiak 's conjugate gaussian models for this.

### CODE REVIEW

File: profiler/distributions.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+def run_with_tool(tool, dists, batch_size):
+    column_widths, field_format, template = None, None, None
+    if tool == 'timeit':
+        field_format = (None, '{:.6f}', '{:.6f}')
+        template = 'column'
+    elif tool == 'cprofile':
+        column_widths = [16, 80]
+        template = 'row'
+    with profile_print(column_widths, field_format, template) as out:
+        out.header('DISTRIBUTION', 'SAMPLE', 'BATCH_LOG_PDF')
```
Comment (fritzo) [inline profiler/distributions.py]: nit: What are the units? Is this ops/sec or ms/op or something else?

File: profiler/distributions.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+        repeat = 5
+        if args.repeat is not None:
+            repeat = args.repeat
+        tool_cfg = {'repeat': repeat}
+    TOOL_CFG = tool_cfg
+
+
+def main():
+    parser = argparse.ArgumentParser(description='Profiling distributions library')
+    parser.add_argument('--tool', nargs='?', default='timeit')
```
Comment (fritzo) [inline profiler/distributions.py]: It would be really nice to have a `--help` option that prints available `--tool`s and `--dist`s.

File: setup.py [hunk 3]
```diff
@@ -36,6 +36,9 @@
             'visdom',
             'torchvision',
         ],
+        'profile': [
```
Comment (fritzo) [inline setup.py]: consider adding `profile` to the `make install` invocation

**Comment by fritzo** [comment]
Could you also add a "Profiling" section at the end of `CONTRIBUTING.md`?

**Comment by martinjankowiak** [comment]
this is cool, nice!

File: profiler/distributions.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+                                lognormal, normal, poisson, uniform)
+
+
+def T(arr):
+    return Variable(torch.Tensor(arr))
+
+
+TOOL = 'timeit'
+TOOL_CFG = {}
+DISTRIBUTIONS = {
```
Comment (jpchen) [inline profiler/distributions.py]: itd be nice to get this from tests/distributions/conf.py

File: profiler/distributions.py (hunk 1 above)
Comment (neerajprad) [reply profiler/distributions.py]: This is just seconds taken for the sampling/scoring operation for the given tensor size. Will make it clearer in the header.

File: profiler/distributions.py (hunk 2 above)
Comment (neerajprad) [reply profiler/distributions.py]: Goot point, I'll add that.

File: setup.py (hunk 3 above)
Comment (neerajprad) [reply setup.py]: Will do.

**Comment by eb8680** [comment]
Cool!

File: profiler/distributions.py (hunk 4 above)
Comment (neerajprad) [reply profiler/distributions.py]: I tried to see if there was some way to share fixtures / test cases, but opted for a more decoupled code in the end. Maybe we can build a pytest plugin that does this too, so that we have all the flexibility of parameterized fixtures, but right now it seemed like taking on much more complexity than what we need.

Comment (fritzo) [reply profiler/distributions.py]: I agree it's cleaner to decouple them, so we can choose examples differently in the two cases: test corner cases vs profile numerically consistent cases.

**Comment by neerajprad** [comment]
> How does the batch stuff work? I often test with tensors of size 10,100,1000,10000,100000 and show in a single table.

I made a change so that this can be varied now. By default this only shows the results for 10,000 and 100,000, so that the output width remains small, but the user can specify different batch sizes now.

```
+--------------+----------------+----------------+----------------+----------------+
| DISTRIBUTION | SAMPLE         | BATCH_LOG_PDF  | SAMPLE         | BATCH_LOG_PDF  |
|              | (N=10000)      | (N=10000)      | (N=100000)     | (N=100000)     |
+--------------+----------------+----------------+----------------+----------------+
| bernoulli    | 0.000577       | 0.000607       | 0.004978       | 0.004455       |
+--------------+----------------+----------------+----------------+----------------+
```

**Comment by neerajprad** [comment]
Good to merge now.
